In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Data Loading

In [3]:
data = pd.read_csv("project_data.csv")
data

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
4            4.0   1.229658   0.141004  0.045371  1.202613  0.191881   
5            7.0  -0.894286   0.286157 -0.113192 -0.271526  2.669599   
6            9.0  -0.338262   1.119593  1.044367 -0.222187  0.499361   
7           10.0   1.249999  -1.221637  0.383930 -1.234899 -1.485419   
8           11.0   1.069374   0.287722  0.828613  2.712520 -0.178398   
9           12.0  -2.791855  -0.327771  1.641750  1.767473 -0.136588   
10          12.0  -0.752417   0.345485  2.057323 -1.468643 -1.158394   
11          13.0  -0.436905   0.918966  0.924591 -0.727219  0.915679   
12          15.0   1.492936  -1.029346  0.454795 -1.438026 -1.555434   
13          16.0   0.694885  -1.361819  1.029221  0.834159 -1.191209   
14          18.0   1.166616   0.502120 -0.067300  2.261569  0.428804   
15          18.0   0.247491   0.277666  1.185471 -0.092603 -1.314394   
16          22.0  -1.946525  -0.044901 -0.405570 -1.013057  2.941968   
17          22.0  -2.074295  -0.121482  1.322021  0.410008  0.295198   
18          23.0   1.173285   0.353498  0.283905  1.133563 -0.172577   
19          23.0   1.322707  -0.174041  0.434555  0.576038 -0.836758   
20          23.0  -0.414289   0.905437  1.727453  1.473471  0.007443   
21          23.0   1.059387  -0.175319  1.266130  1.186110 -0.786002   
22          24.0   1.237429   0.061043  0.380526  0.761564 -0.359771   
23          25.0   1.114009   0.085546  0.493702  1.335760 -0.300189   
24          26.0  -0.529912   0.873892  1.347247  0.145457  0.414209   
25          26.0  -0.535388   0.865268  1.351076  0.147575  0.433680   
26          27.0  -1.452187   1.765124  0.611669  1.176825 -0.445980   
27          32.0   1.249055  -0.624727 -0.710589 -0.991600  1.429973   
28          32.0  -2.008872   2.198527  0.144242  1.159432 -0.815174   
29          33.0  -0.607877   1.031345  1.740450  1.232106  0.418592   
...          ...        ...        ...       ...       ...       ...   
199334  172758.0  -0.546378   1.433992 -0.313252  0.926044  0.522388   
199335  172759.0  -1.570301   1.880590 -0.559004 -1.003017 -0.326096   
199336  172759.0   1.894910  -0.427906 -2.184267  0.159979  0.587740   
199337  172760.0  -6.713826   2.369104 -3.534768 -0.368259 -1.721030   
199338  172761.0  -0.066136   0.996296 -0.918564 -0.923242  1.256522   
199339  172762.0   1.955547  -0.724606 -1.706511 -0.611145  1.710907   
199340  172762.0  -1.143909   1.780421 -0.454124 -0.814393  0.514689   
199341  172764.0  -0.764523   0.588379 -0.907599 -0.418847  0.901528   
199342  172766.0   1.975178  -0.616244 -2.628295 -0.406246  2.327804   
199343  172766.0  -1.139015  -0.155510  1.894478 -1.138957  1.451777   
199344  172767.0  -0.268061   2.540315 -1.400915  4.846661  0.639105   
199345  172768.0  -1.796092   1.929178 -2.828417 -1.689844  2.199572   
199346  172768.0  -0.669662   0.923769 -1.543167 -1.560729  2.833960   
199347  172768.0  -2.076175   2.142238 -2.522704 -1.888063  1.982785   
199348  172769.0  -1.029719  -1.110670 -0.636179 -0.840816  2.424360   
199349  172770.0   2.007418  -0.280235 -0.208113  0.335261 -0.715798   
199350  172770.0  -0.446951   1.302212 -0.168583  0.981577  0.578957   
199351  172771.0  -0.515513   0.971950 -1.014580 -0.677037  0.912430   
199352  172774.0  -0.863506   0.874701  0.420358 -0.530365  0.356561   
199353  172774.0  -0.724123   1.485216 -1.132218 -0.607190  0.709499   
199354  172775.0   1.971002  -0.699067 -1.697541 -0.617643  1.718797   
199355  172777.0  -1.266580  -0.400461  0.956221 -0.723919  1.531993   
199356  172778.0 -12.516732  10.187818 -8.476671 -2.510473 -4.586669   
199357  172780.0   1.884849  -0.143540 -0.999943  1.

### 원하는 feature 뽑아오기

In [4]:
def SelectFeature(data, feature_list):
    
    X = data[feature_list].values
    y = data['Class'].values
    
    return X,y;

### 테스트 케이스 나누기

In [5]:
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(data, test_size=0.2, random_state=42)


### f1 Measure

In [6]:
from sklearn.metrics import f1_score

def CalScore(true,predict):

    
    f1Score = f1_score(true, predict, average='binary')
    print(f1Score)
    '''
    f1Score = f1_score(true, predict, average='macro')
    print(f1Score)
    f1Score = f1_score(true, predict, average='micro')
    print(f1Score)
    f1Score = f1_score(true, predict, average='weighted')
    print(f1Score)
    f1Score = f1_score(true, predict, average=None)
    print(f1Score)
    '''

### f beta Measure

In [7]:
from sklearn.metrics import fbeta_score

def CalBetaScore(true,predict):
    fBetaScore = fbeta_score(true, predict, average='binary', beta=1.5) #recall을 더 높은 비중으로
    print(fBetaScore)

# Feature Selection

우선 time에 따른 feature값을 비교

<img src = "img/feature_time.png">

1. feature 전체 
2. feature들 중 abnormal이 극명하게 들어나는 V11,V12,V14,V16,V17을 선택 (feature 5개)
3. Class값과 Relation이 높은 V1, V2, V3, V4, V5, V7, V9, V10, V11, V12, V14, V16, V17, V18을 선택 (feature 14개)

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

#feature 전체
feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

svm = SVC(kernel='rbf')
svm.fit(train_X, train_y)

predict = svm.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

#좋은 feature 선별
feature_sub = ['V11','V12','V14','V16','V17'];

train_X,train_y = SelectFeature(train_set,feature_sub);
test_X , test_y = SelectFeature(test_set,feature_sub);

svm = SVC(kernel='rbf')
svm.fit(train_X, train_y)

predict = svm.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

#Correlation이 높은 feature 선별
feature_sub2 = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_sub2);
test_X , test_y = SelectFeature(test_set,feature_sub2);

svm = SVC(kernel='rbf')
svm.fit(train_X, train_y)

predict = svm.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

C:\Users\opp06\Anaconda3\envs\MLProject\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.736
0.6795454545454545


C:\Users\opp06\Anaconda3\envs\MLProject\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.8270676691729324
0.7839912280701753


C:\Users\opp06\Anaconda3\envs\MLProject\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.6495726495726496
0.5825471698113207


# Model Selection

In [30]:
feature_list = ['V11','V12','V14','V16','V17'];

train_X,train_y = SelectFeature(train_set,feature_list);
test_X , test_y = SelectFeature(test_set,feature_list);

다양한 Classification 알고리즘 비교 
- Logistic Regression 
- SVM 
- Decision Tree 
- Random Forest 
- Gradient Boosting Classifier 
- MLP 
- k-NN 

### Logistic Regression

In [31]:
from sklearn.linear_model import LogisticRegression

logit = LogisticRegression()
logit.fit(train_X, train_y)

#predict = logit.predict(train_X)
#CalScore(train_y,predict)

predict = logit.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

C:\Users\opp06\Anaconda3\envs\MLProject\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.6451612903225806
0.5936073059360731


### SVM

In [13]:
from sklearn.svm import SVC

svm = SVC(kernel='rbf')
svm.fit(train_X, train_y)

#predict = svm.predict(train_X)
#CalScore(train_y,predict)

predict = svm.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

C:\Users\opp06\Anaconda3\lib\site-packages\sklearn\svm\base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.8270676691729324
0.7839912280701753


### Decision Tree Classifier

In [19]:
from sklearn import tree

decisionTree = tree.DecisionTreeClassifier()
decisionTree.fit(train_X, train_y)

predict = decisionTree.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

0.7297297297297298
0.7222222222222222


### Random Forest

In [20]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

rfc = RandomForestClassifier(random_state = 0, n_estimators = 100,criterion = 'entropy')
rfc.fit(train_X, train_y)

#predict = rfc.predict(train_X)
#CalScore(train_y,predict)

predict = rfc.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

0.842857142857143
0.8159574468085106


### Gradient Boosting Classifier

In [21]:
from sklearn.ensemble import GradientBoostingClassifier

gbc = GradientBoostingClassifier()
gbc.fit(train_X, train_y)

predict = gbc.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)


0.7194244604316548
0.6944444444444444


In [12]:
import xgboost as xgb

xgb_model = xgb.XGBClassifier()
xgb_model.fit(train_X, train_y)

predict = xgb_model.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

0.7943262411347517
0.771186440677966


### MLP Classifier

In [13]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1)  
mlp.fit(train_X, train_y)  

predict = mlp.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

0.786206896551724
0.771875


### k-NN

In [14]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
knn.fit(train_X, train_y)

predict = knn.predict(test_X)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)

0.8201438848920864
0.7916666666666669


모델 학습값 평가
<img src="img/model_select.PNG">

### svm 파라미터 튜닝 with GridSearchCV

In [19]:
from sklearn.model_selection import GridSearchCV

param_grid = {'C': [0.001,0.01,0.1, 1, 10, 100,1000], 'gamma': ['auto', 1, 0.1, 0.01, 0.001, 0.00001]}


clf = GridSearchCV(SVC(), param_grid, n_jobs=-1,verbose=1, scoring='f1')
clf.fit(train_X, train_y)

print("Best Parameters:\n", clf.best_params_)
print("Best Estimators:\n", clf.best_estimator_)

d:\python\python37\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 16 concurrent workers.


Fitting 3 folds for each of 42 candidates, totalling 126 fits


[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:    9.8s
[Parallel(n_jobs=-1)]: Done 126 out of 126 | elapsed:  5.2min finished


Best Parameters:
 {'C': 10, 'gamma': 0.1}
Best Estimators:
 SVC(C=10, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.1, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


In [20]:
from sklearn.svm import SVC

svm =   SVC(C=10, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.1, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
svm.fit(train_X, train_y)

#predict = svm.predict(train_X)
#CalScore(train_y,predict)

predict2 = svm.predict(test_X)
CalScore(test_y,predict2)

0.7999999999999999


f1을 score로 했음에도 불구하고 성능 향상이 되지 않음

In [7]:
valueArray = [];

for i in range(28):
    valueArray.append(data['V'+str(i+1)])

In [8]:
col = np.where(result==0,'b','r')

for i in range(len(valueArray)):
    for j in range(len(valueArray)):
        print(i+1,j+1)
        #plt.scatter(valueArray[i],valueArray[j],c=col,s=10)
        #plt.savefig("cov"+str(i)+str(j)+".png")
        #plt.clf()
        

1 1
1 2
1 3
1 4
1 5
1 6
1 7
1 8
1 9
1 10
1 11
1 12
1 13
1 14
1 15
1 16
1 17
1 18
1 19
1 20
1 21
1 22
1 23
1 24
1 25
1 26
1 27
1 28
2 1
2 2
2 3
2 4
2 5
2 6
2 7
2 8
2 9
2 10
2 11
2 12
2 13
2 14
2 15
2 16
2 17
2 18
2 19
2 20
2 21
2 22
2 23
2 24
2 25
2 26
2 27
2 28
3 1
3 2
3 3
3 4
3 5
3 6
3 7
3 8
3 9
3 10
3 11
3 12
3 13
3 14
3 15
3 16
3 17
3 18
3 19
3 20
3 21
3 22
3 23
3 24
3 25
3 26
3 27
3 28
4 1
4 2
4 3
4 4
4 5
4 6
4 7
4 8
4 9
4 10
4 11
4 12
4 13
4 14
4 15
4 16
4 17
4 18
4 19
4 20
4 21
4 22
4 23
4 24
4 25
4 26
4 27
4 28
5 1
5 2
5 3
5 4
5 5
5 6
5 7
5 8
5 9
5 10
5 11
5 12
5 13
5 14
5 15
5 16
5 17
5 18
5 19
5 20
5 21
5 22
5 23
5 24
5 25
5 26
5 27
5 28
6 1
6 2
6 3
6 4
6 5
6 6
6 7
6 8
6 9
6 10
6 11
6 12
6 13
6 14
6 15
6 16
6 17
6 18
6 19
6 20
6 21
6 22
6 23
6 24
6 25
6 26
6 27
6 28
7 1
7 2
7 3
7 4
7 5
7 6
7 7
7 8
7 9
7 10
7 11
7 12
7 13
7 14
7 15
7 16
7 17
7 18
7 19
7 20
7 21
7 22
7 23
7 24
7 25
7 26
7 27
7 28
8 1
8 2
8 3
8 4
8 5
8 6
8 7
8 8
8 9
8 10
8 11
8 12
8 13
8 14
8 15
8 16
8 17
8 18
8 

In [9]:

for i in range(len(valueArray)):
    print(i)
    print(np.mean(valueArray[i]))
    print(np.var(valueArray[i]))
    print(np.std(valueArray[i]))
    #plt.scatter(time,valueArray[i],c=col,s=10)
    #plt.savefig("value"+str(i)+".png")
    #plt.clf()

0
-0.003370153263307601
3.849591828427786
1.9620376725302158
1
0.0009989272770197716
2.757126388159257
1.6604596918200867
2
0.0004080125078719491
2.3041277610233655
1.5179353612797106
3
0.0031398879531128274
2.010639537183285
1.417970217311804
4
0.0006059713480719164
1.9214211708025348
1.3861533720344712
5
-0.00016432192844623237
1.7835095724498509
1.335481026615448
6
0.001902354583129389
1.5576292337595443
1.2480501727733322
7
-0.0011247242026774546
1.4312386917702324
1.1963438852479802
8
0.0007226963880086407
1.2094352889259228
1.0997432831920015
9
-0.001475384838393684
1.1971786561847864
1.0941565958238273
10
0.0013898046105778248
1.0444270844230639
1.0219721544264617
11
-0.0016884974514489828
1.0072842222987084
1.0036355027093793
12
-0.0031960345874183127
0.9910193103362939
0.9954995280442346
13
-0.0009140242950301451
0.9249473483552314
0.9617418304073247
14
-0.00028474145381528735
0.8404485267287229
0.9167597977271489
15
0.000534280165461583
0.7698000485081136
0.8773824984054067
1

# Neural Network

일반적인 알고리즘의 한계를 느끼고 Nerural Network를 사용

In [70]:
# Importing the Keras libraries and packages
from tensorflow import keras
from tensorflow.keras import layers
# Initialising the ANN
classifier = keras.Sequential()
# Adding the input layer and the first hidden layer
classifier.add(layers.Dense(units =15 , kernel_initializer = 'uniform', activation = 'relu', input_dim = 5))
# Adding the second hidden layer
classifier.add(layers.Dense(units = 15, kernel_initializer = 'uniform', activation = 'relu'))
# Adding the output layer
classifier.add(layers.Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
# Compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
# Fitting the ANN to the Training set
classifier.fit(train_X, train_y, batch_size = 32, epochs = 10)

Epoch 1/10
159491/159491 [==============================] - 20s 123us/sample - loss: 0.0207 - acc: 0.9984
Epoch 2/10
159491/159491 [==============================] - 19s 119us/sample - loss: 0.0039 - acc: 0.9993
Epoch 3/10
159491/159491 [==============================] - 19s 119us/sample - loss: 0.0038 - acc: 0.9993
Epoch 4/10
159491/159491 [==============================] - 20s 122us/sample - loss: 0.0037 - acc: 0.9993
Epoch 5/10
159491/159491 [==============================] - 19s 121us/sample - loss: 0.0037 - acc: 0.9993
Epoch 6/10
159491/159491 [==============================] - 19s 121us/sample - loss: 0.0037 - acc: 0.9993
Epoch 7/10
159491/159491 [==============================] - 19s 122us/sample - loss: 0.0036 - acc: 0.9994
Epoch 8/10
159491/159491 [==============================] - 20s 123us/sample - loss: 0.0036 - acc: 0.9994
Epoch 9/10
159491/159491 [==============================] - 19s 122us/sample - loss: 0.0036 - acc: 0.9993
Epoch 10/10
159491/159491 [===================

In [74]:
# Predicting the Test set results
y_pred = classifier.predict(test_X)
y_pred = (y_pred > 0.5)
print(y_pred)
score = classifier.evaluate(test_X, test_y)
score

[[False]
 [False]
 [False]
 ...
 [False]
 [False]
 [False]]
39873/39873 [==============================] - 2s 44us/sample - loss: 0.0042 - acc: 0.9992


[0.004205411993305984, 0.9992476]

In [73]:
#Let's see how our model performed
from sklearn.metrics import classification_report
print(classification_report(test_y, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     39797
           1       0.83      0.76      0.79        76

    accuracy                           1.00     39873
   macro avg       0.91      0.88      0.90     39873
weighted avg       1.00      1.00      1.00     39873



In [136]:

#feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

#feature 전체
feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];


train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

In [137]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

model = tf.keras.Sequential()

model.add(layers.Dense(units=32, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = 28))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=10,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/10
159491/159491 [==============================] - 19s 121us/sample - loss: 0.0130 - acc: 0.9978
Epoch 2/10
159491/159491 [==============================] - 19s 120us/sample - loss: 0.0031 - acc: 0.9994
Epoch 3/10
159491/159491 [==============================] - 19s 120us/sample - loss: 0.0027 - acc: 0.9994
Epoch 4/10
159491/159491 [==============================] - 19s 122us/sample - loss: 0.0026 - acc: 0.9994
Epoch 5/10
159491/159491 [==============================] - 19s 121us/sample - loss: 0.0027 - acc: 0.9994
Epoch 6/10
159491/159491 [==============================] - 19s 121us/sample - loss: 0.0023 - acc: 0.9995
Epoch 7/10
159491/159491 [==============================] - 19s 122us/sample - loss: 0.0023 - acc: 0.9995
Epoch 8/10
159491/159491 [==============================] - 19s 121us/sample - loss: 0.0021 - acc: 0.9995
Epoch 9/10
159491/159491 [==============================] - 19s 119us/sample - loss: 0.0020 - acc: 0.9994
Epoch 10/10
159491/159491 [===================

In [138]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

model = tf.keras.Sequential()
model.add(layers.Dense(units=32, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = 28))

model.add(layers.Dense(units=32,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=10,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/10
159491/159491 [==============================] - 23s 141us/sample - loss: 0.0120 - acc: 0.9968
Epoch 2/10
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0031 - acc: 0.9994
Epoch 3/10
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0028 - acc: 0.9994
Epoch 4/10
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0025 - acc: 0.9994
Epoch 5/10
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0023 - acc: 0.9995
Epoch 6/10
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0023 - acc: 0.9995
Epoch 7/10
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0021 - acc: 0.9995
Epoch 8/10
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0019 - acc: 0.9995
Epoch 9/10
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0018 - acc: 0.9995
Epoch 10/10
159491/159491 [===================

In [139]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

model = tf.keras.Sequential()
model.add(layers.Dense(units=32, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = 28))

model.add(layers.Dense(units=32,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=32,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=10,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/10
159491/159491 [==============================] - 26s 161us/sample - loss: 0.0073 - acc: 0.9990
Epoch 2/10
159491/159491 [==============================] - 25s 159us/sample - loss: 0.0031 - acc: 0.9993
Epoch 3/10
159491/159491 [==============================] - 25s 158us/sample - loss: 0.0029 - acc: 0.9994
Epoch 4/10
159491/159491 [==============================] - 26s 162us/sample - loss: 0.0026 - acc: 0.9994
Epoch 5/10
159491/159491 [==============================] - 25s 159us/sample - loss: 0.0024 - acc: 0.9994
Epoch 6/10
159491/159491 [==============================] - 25s 160us/sample - loss: 0.0024 - acc: 0.9994
Epoch 7/10
159491/159491 [==============================] - 25s 158us/sample - loss: 0.0021 - acc: 0.9995
Epoch 8/10
159491/159491 [==============================] - 25s 159us/sample - loss: 0.0020 - acc: 0.9995
Epoch 9/10
159491/159491 [==============================] - 25s 159us/sample - loss: 0.0020 - acc: 0.9995
Epoch 10/10
159491/159491 [===================

In [140]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

model = tf.keras.Sequential()
model.add(layers.Dense(units=32, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = 28))

model.add(layers.Dense(units=32,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=32,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=32,
                        kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                        activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=10,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/10
159491/159491 [==============================] - 29s 182us/sample - loss: 0.0075 - acc: 0.9991
Epoch 2/10
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0033 - acc: 0.9994
Epoch 3/10
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0030 - acc: 0.9994
Epoch 4/10
159491/159491 [==============================] - 28s 179us/sample - loss: 0.0028 - acc: 0.9994
Epoch 5/10
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0025 - acc: 0.9994
Epoch 6/10
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0023 - acc: 0.9994
Epoch 7/10
159491/159491 [==============================] - 29s 181us/sample - loss: 0.0021 - acc: 0.9994
Epoch 8/10
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0021 - acc: 0.9995
Epoch 9/10
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0021 - acc: 0.9994
Epoch 10/10
159491/159491 [===================

Hidden Layer 3개일 때 최적의 결과가 나옴을 확인

Hidden Layer의 유닛과 feature를 비교

<img src="img/ml_model_select.PNG">

In [149]:

feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 14

model = tf.keras.Sequential()
model.add(layers.Dense(units=hidden_unit, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = feature_count))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=10,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/10
159491/159491 [==============================] - 26s 164us/sample - loss: 0.0163 - acc: 0.9969
Epoch 2/10
159491/159491 [==============================] - 26s 162us/sample - loss: 0.0035 - acc: 0.9994
Epoch 3/10
159491/159491 [==============================] - 26s 163us/sample - loss: 0.0032 - acc: 0.9994
Epoch 4/10
159491/159491 [==============================] - 26s 164us/sample - loss: 0.0031 - acc: 0.9993
Epoch 5/10
159491/159491 [==============================] - 26s 161us/sample - loss: 0.0030 - acc: 0.9994
Epoch 6/10
159491/159491 [==============================] - 26s 163us/sample - loss: 0.0029 - acc: 0.9994
Epoch 7/10
159491/159491 [==============================] - 25s 159us/sample - loss: 0.0028 - acc: 0.9994
Epoch 8/10
159491/159491 [==============================] - 26s 161us/sample - loss: 0.0027 - acc: 0.9994
Epoch 9/10
159491/159491 [==============================] - 26s 163us/sample - loss: 0.0027 - acc: 0.9994
Epoch 10/10
159491/159491 [===================

In [158]:

feature_total = feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 28

model = tf.keras.Sequential()
model.add(layers.Dense(units=hidden_unit, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = feature_count))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=10,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/10
159491/159491 [==============================] - 28s 178us/sample - loss: 0.0097 - acc: 0.9979
Epoch 2/10
159491/159491 [==============================] - 28s 173us/sample - loss: 0.0032 - acc: 0.9994
Epoch 3/10
159491/159491 [==============================] - ETA: 0s - loss: 0.0029 - acc: 0.999 - 28s 176us/sample - loss: 0.0029 - acc: 0.9994
Epoch 4/10
159491/159491 [==============================] - 28s 173us/sample - loss: 0.0026 - acc: 0.9994
Epoch 5/10
159491/159491 [==============================] - 27s 172us/sample - loss: 0.0024 - acc: 0.9994
Epoch 6/10
159491/159491 [==============================] - 28s 175us/sample - loss: 0.0022 - acc: 0.9995
Epoch 7/10
159491/159491 [==============================] - 28s 176us/sample - loss: 0.0022 - acc: 0.9994
Epoch 8/10
159491/159491 [==============================] - 28s 177us/sample - loss: 0.0020 - acc: 0.9995
Epoch 9/10
159491/159491 [==============================] - 28s 173us/sample - loss: 0.0019 - acc: 0.9995
Epoch 10

유닛이 28개일때, feature14, feature total에서 높은 수치가 나옴

epoch수를 늘려서 모델의 정확성을 높힘

<img src="img/ml_epoch.PNG">

epoch 100 for feature14, feature total

In [164]:

feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 14

model = tf.keras.Sequential()
model.add(layers.Dense(units=hidden_unit, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = feature_count))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=100,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0114 - acc: 0.9987
Epoch 2/100
159491/159491 [==============================] - 28s 177us/sample - loss: 0.0033 - acc: 0.9994
Epoch 3/100
159491/159491 [==============================] - 29s 179us/sample - loss: 0.0031 - acc: 0.9994
Epoch 4/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0030 - acc: 0.9994
Epoch 5/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0029 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 29s 182us/sample - loss: 0.0028 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0027 - acc: 0.9994
Epoch 8/100
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0027 - acc: 0.9994
Epoch 9/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0026 - acc: 0.9994
Epoch 10/100
159491/159491 [=========

In [165]:
feature_total = feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 28

model = tf.keras.Sequential()
model.add(layers.Dense(units=hidden_unit, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = feature_count))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=100,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0075 - acc: 0.9987
Epoch 2/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0032 - acc: 0.9993
Epoch 3/100
159491/159491 [==============================] - 29s 180us/sample - loss: 0.0028 - acc: 0.9994
Epoch 4/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0025 - acc: 0.9994
Epoch 5/100
159491/159491 [==============================] - 29s 182us/sample - loss: 0.0023 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0022 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 29s 183us/sample - loss: 0.0020 - acc: 0.9995
Epoch 8/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0019 - acc: 0.9995
Epoch 9/100
159491/159491 [==============================] - 29s 182us/sample - loss: 0.0018 - acc: 0.9995
Epoch 10/100
159491/159491 [=========

159491/159491 [==============================] - 29s 184us/sample - loss: 5.3802e-04 - acc: 0.9998
Epoch 76/100
159491/159491 [==============================] - 29s 185us/sample - loss: 4.9579e-04 - acc: 0.9999
Epoch 77/100
159491/159491 [==============================] - 29s 179us/sample - loss: 5.0656e-04 - acc: 0.9998
Epoch 78/100
159491/159491 [==============================] - 29s 182us/sample - loss: 5.0795e-04 - acc: 0.9999
Epoch 79/100
159491/159491 [==============================] - 29s 183us/sample - loss: 5.2621e-04 - acc: 0.9999
Epoch 80/100
159491/159491 [==============================] - 29s 182us/sample - loss: 4.1094e-04 - acc: 0.9999
Epoch 81/100
159491/159491 [==============================] - 29s 184us/sample - loss: 3.8030e-04 - acc: 0.9999
Epoch 82/100
159491/159491 [==============================] - 29s 183us/sample - loss: 6.1484e-04 - acc: 0.9998
Epoch 83/100
159491/159491 [==============================] - 29s 184us/sample - loss: 3.2855e-04 - acc: 0.9999
Epoch

epoch 200 for feature14, feature total

In [9]:

feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 14

model = tf.keras.Sequential()
model.add(layers.Dense(units=hidden_unit, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = feature_count))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=200,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/200
159491/159491 [==============================] - 23s 143us/sample - loss: 0.0126 - acc: 0.9984
Epoch 2/200
159491/159491 [==============================] - 23s 144us/sample - loss: 0.0034 - acc: 0.9994
Epoch 3/200
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0033 - acc: 0.9994
Epoch 4/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0031 - acc: 0.9994
Epoch 5/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0030 - acc: 0.9994
Epoch 6/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0030 - acc: 0.9994
Epoch 7/200
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0028 - acc: 0.9994
Epoch 8/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0028 - acc: 0.9994
Epoch 9/200
159491/159491 [==============================] - 23s 142us/sample - loss: 0.0028 - acc: 0.9994
Epoch 10/200
159491/159491 [=========

159491/159491 [==============================] - 23s 143us/sample - loss: 0.0012 - acc: 0.9996
Epoch 77/200
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0011 - acc: 0.9997
Epoch 78/200
159491/159491 [==============================] - 22s 141us/sample - loss: 0.0010 - acc: 0.9997
Epoch 79/200
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0012 - acc: 0.9996
Epoch 80/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0013 - acc: 0.9996
Epoch 81/200
159491/159491 [==============================] - 22s 138us/sample - loss: 0.0010 - acc: 0.9997
Epoch 82/200
159491/159491 [==============================] - 23s 142us/sample - loss: 0.0011 - acc: 0.9997
Epoch 83/200
159491/159491 [==============================] - 22s 141us/sample - loss: 0.0011 - acc: 0.9997
Epoch 84/200
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0011 - acc: 0.9997
Epoch 85/200
159491/159491 [=============

Epoch 150/200
159491/159491 [==============================] - 22s 141us/sample - loss: 6.9494e-04 - acc: 0.9997
Epoch 151/200
159491/159491 [==============================] - 22s 139us/sample - loss: 7.7329e-04 - acc: 0.9998
Epoch 152/200
159491/159491 [==============================] - 22s 139us/sample - loss: 7.4596e-04 - acc: 0.9998
Epoch 153/200
159491/159491 [==============================] - 22s 137us/sample - loss: 9.2535e-04 - acc: 0.9997
Epoch 154/200
159491/159491 [==============================] - 22s 137us/sample - loss: 7.8176e-04 - acc: 0.9998
Epoch 155/200
159491/159491 [==============================] - 22s 139us/sample - loss: 9.3652e-04 - acc: 0.9997
Epoch 156/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0012 - acc: 0.9997
Epoch 157/200
159491/159491 [==============================] - 22s 141us/sample - loss: 0.0010 - acc: 0.9997
Epoch 158/200
159491/159491 [==============================] - 23s 143us/sample - loss: 6.8383e-04 - acc

In [10]:
feature_total = feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 28

model = tf.keras.Sequential()
model.add(layers.Dense(units=hidden_unit, 
                       kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                       activation='relu',
                       input_dim = feature_count))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=hidden_unit,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='relu'))

model.add(layers.Dense(units=1,
                       kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                       activation='sigmoid'))


model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(train_X, train_y,epochs=200,batch_size=32)


#검증
predict = model.predict(test_X)
predict = (predict > 0.5)
from sklearn.metrics import classification_report
print(classification_report(test_y, predict))


predict_y = predict.reshape(-1)
CalScore(test_y,predict_y)
CalBetaScore(test_y,predict_y)


Epoch 1/200
159491/159491 [==============================] - 23s 141us/sample - loss: 0.0105 - acc: 0.9980
Epoch 2/200
159491/159491 [==============================] - 22s 141us/sample - loss: 0.0033 - acc: 0.9993
Epoch 3/200
159491/159491 [==============================] - 22s 141us/sample - loss: 0.0028 - acc: 0.9994
Epoch 4/200
159491/159491 [==============================] - 22s 135us/sample - loss: 0.0026 - acc: 0.9994
Epoch 5/200
159491/159491 [==============================] - 23s 142us/sample - loss: 0.0023 - acc: 0.9995
Epoch 6/200
159491/159491 [==============================] - 23s 143us/sample - loss: 0.0021 - acc: 0.9995
Epoch 7/200
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0020 - acc: 0.9994
Epoch 8/200
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0019 - acc: 0.9994
Epoch 9/200
159491/159491 [==============================] - 23s 141us/sample - loss: 0.0019 - acc: 0.9995
Epoch 10/200
159491/159491 [=========

159491/159491 [==============================] - 23s 144us/sample - loss: 4.4127e-04 - acc: 0.9999
Epoch 76/200
159491/159491 [==============================] - 23s 143us/sample - loss: 5.7005e-04 - acc: 0.9998
Epoch 77/200
159491/159491 [==============================] - 23s 142us/sample - loss: 5.0108e-04 - acc: 0.9998
Epoch 78/200
159491/159491 [==============================] - 23s 142us/sample - loss: 4.5860e-04 - acc: 0.9999
Epoch 79/200
159491/159491 [==============================] - 23s 141us/sample - loss: 4.6182e-04 - acc: 0.9998
Epoch 80/200
159491/159491 [==============================] - 23s 141us/sample - loss: 3.4312e-04 - acc: 0.9999
Epoch 81/200
159491/159491 [==============================] - 23s 142us/sample - loss: 7.7300e-04 - acc: 0.9998
Epoch 82/200
159491/159491 [==============================] - 23s 145us/sample - loss: 6.6980e-04 - acc: 0.9998
Epoch 83/200
159491/159491 [==============================] - 23s 143us/sample - loss: 3.4913e-04 - acc: 0.9999
Epoch

159491/159491 [==============================] - 23s 142us/sample - loss: 4.9900e-04 - acc: 0.9999
Epoch 148/200
159491/159491 [==============================] - 23s 143us/sample - loss: 6.7992e-04 - acc: 0.9999
Epoch 149/200
159491/159491 [==============================] - 23s 142us/sample - loss: 5.7063e-04 - acc: 0.9999
Epoch 150/200
159491/159491 [==============================] - 22s 141us/sample - loss: 5.4482e-04 - acc: 0.9999
Epoch 151/200
159491/159491 [==============================] - 23s 146us/sample - loss: 3.9908e-04 - acc: 0.9999
Epoch 152/200
159491/159491 [==============================] - 23s 144us/sample - loss: 2.7035e-04 - acc: 0.9999
Epoch 153/200
159491/159491 [==============================] - 23s 144us/sample - loss: 4.8279e-04 - acc: 0.9999
Epoch 154/200
159491/159491 [==============================] - 23s 142us/sample - loss: 5.1485e-04 - acc: 0.9999
Epoch 155/200
159491/159491 [==============================] - 23s 143us/sample - loss: 4.5684e-04 - acc: 0.99

Model Average(Ensemble)

In [166]:
feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 14

n_members = 10
for i in range(n_members):
    model = tf.keras.Sequential()
    model.add(layers.Dense(units=hidden_unit, 
                           kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                           activation='relu',
                           input_dim = feature_count))

    model.add(layers.Dense(units=hidden_unit,
                           kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                           activation='relu'))

    model.add(layers.Dense(units=hidden_unit,
                           kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                           activation='relu'))

    model.add(layers.Dense(units=1,
                           kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                           activation='sigmoid'))


    model.compile(optimizer='adam',
                  loss='binary_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_X, train_y,epochs=100,batch_size=32)
    
    filename = 'model/model14x28x28x28x1_'+ str(i + 1)+ '.h5'
    
    model.save(filename)

Epoch 1/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0099 - acc: 0.9975
Epoch 2/100
159491/159491 [==============================] - 29s 185us/sample - loss: 0.0034 - acc: 0.9994
Epoch 3/100
159491/159491 [==============================] - 29s 182us/sample - loss: 0.0032 - acc: 0.9994
Epoch 4/100
159491/159491 [==============================] - 29s 182us/sample - loss: 0.0030 - acc: 0.9994
Epoch 5/100
159491/159491 [==============================] - 29s 184us/sample - loss: 0.0029 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0027 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 29s 184us/sample - loss: 0.0027 - acc: 0.9994
Epoch 8/100
159491/159491 [==============================] - 29s 184us/sample - loss: 0.0026 - acc: 0.9994
Epoch 9/100
159491/159491 [==============================] - 29s 184us/sample - loss: 0.0025 - acc: 0.9994
Epoch 10/100
159491/159491 [=========

159491/159491 [==============================] - 30s 187us/sample - loss: 0.0013 - acc: 0.9996
Epoch 53/100
159491/159491 [==============================] - 30s 186us/sample - loss: 0.0013 - acc: 0.9996
Epoch 54/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0013 - acc: 0.9996
Epoch 55/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0014 - acc: 0.9996
Epoch 56/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0014 - acc: 0.9996
Epoch 57/100
159491/159491 [==============================] - 29s 184us/sample - loss: 0.0011 - acc: 0.9996
Epoch 58/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0014 - acc: 0.9996
Epoch 59/100
159491/159491 [==============================] - 30s 185us/sample - loss: 0.0011 - acc: 0.9996
Epoch 60/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0012 - acc: 0.9996
Epoch 61/100
159491/159491 [=============

Epoch 3/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0034 - acc: 0.9994
Epoch 4/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0031 - acc: 0.9994
Epoch 5/100
159491/159491 [==============================] - 30s 187us/sample - loss: 0.0030 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 30s 187us/sample - loss: 0.0029 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 30s 187us/sample - loss: 0.0029 - acc: 0.9994
Epoch 8/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0027 - acc: 0.9995
Epoch 9/100
159491/159491 [==============================] - 30s 187us/sample - loss: 0.0028 - acc: 0.9994
Epoch 10/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0026 - acc: 0.9994
Epoch 11/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0026 - acc: 0.9994
Epoch 12/100
159491/159491 [=======

159491/159491 [==============================] - 30s 189us/sample - loss: 0.0012 - acc: 0.9996
Epoch 54/100
159491/159491 [==============================] - 30s 190us/sample - loss: 0.0014 - acc: 0.9996
Epoch 55/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0013 - acc: 0.9996
Epoch 56/100
159491/159491 [==============================] - 30s 190us/sample - loss: 0.0014 - acc: 0.9996
Epoch 57/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0014 - acc: 0.9996
Epoch 58/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0013 - acc: 0.9996
Epoch 59/100
159491/159491 [==============================] - 30s 188us/sample - loss: 0.0012 - acc: 0.9997
Epoch 60/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0012 - acc: 0.9996
Epoch 61/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0012 - acc: 0.9996
Epoch 62/100
159491/159491 [=============

159491/159491 [==============================] - 31s 192us/sample - loss: 0.0031 - acc: 0.9994
Epoch 4/100
159491/159491 [==============================] - 30s 190us/sample - loss: 0.0031 - acc: 0.9994
Epoch 5/100
159491/159491 [==============================] - 30s 190us/sample - loss: 0.0030 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0029 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0028 - acc: 0.9994
Epoch 8/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0027 - acc: 0.9994
Epoch 9/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0026 - acc: 0.9994
Epoch 10/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0026 - acc: 0.9994
Epoch 11/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0025 - acc: 0.9994
Epoch 12/100
159491/159491 [===================

159491/159491 [==============================] - 30s 191us/sample - loss: 0.0012 - acc: 0.9996
Epoch 80/100
159491/159491 [==============================] - 30s 189us/sample - loss: 0.0012 - acc: 0.9996
Epoch 81/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0012 - acc: 0.9995
Epoch 82/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0011 - acc: 0.9996
Epoch 83/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0012 - acc: 0.9996
Epoch 84/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0012 - acc: 0.9996
Epoch 85/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0012 - acc: 0.9996
Epoch 86/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0014 - acc: 0.9995
Epoch 87/100
159491/159491 [==============================] - 30s 191us/sample - loss: 0.0012 - acc: 0.9996
Epoch 88/100
159491/159491 [=============

159491/159491 [==============================] - 31s 192us/sample - loss: 0.0014 - acc: 0.9996
Epoch 56/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0014 - acc: 0.9996
Epoch 57/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0015 - acc: 0.9996
Epoch 58/100
159491/159491 [==============================] - 31s 193us/sample - loss: 0.0014 - acc: 0.9996
Epoch 59/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0015 - acc: 0.9996
Epoch 60/100
159491/159491 [==============================] - 31s 193us/sample - loss: 0.0013 - acc: 0.9996
Epoch 61/100
159491/159491 [==============================] - 31s 192us/sample - loss: 0.0015 - acc: 0.9996
Epoch 62/100
159491/159491 [==============================] - 31s 194us/sample - loss: 0.0012 - acc: 0.9996
Epoch 63/100
159491/159491 [==============================] - 31s 193us/sample - loss: 0.0013 - acc: 0.9996
Epoch 64/100
159491/159491 [=============

159491/159491 [==============================] - 31s 194us/sample - loss: 0.0029 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0028 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 31s 194us/sample - loss: 0.0028 - acc: 0.9994
Epoch 8/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0027 - acc: 0.9994
Epoch 9/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0027 - acc: 0.9994
Epoch 10/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0026 - acc: 0.9994
Epoch 11/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0026 - acc: 0.9994
Epoch 12/100
159491/159491 [==============================] - 31s 197us/sample - loss: 0.0026 - acc: 0.9994
Epoch 13/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0025 - acc: 0.9995
Epoch 14/100
159491/159491 [=================

159491/159491 [==============================] - 31s 196us/sample - loss: 0.0013 - acc: 0.9996
Epoch 82/100
159491/159491 [==============================] - 31s 196us/sample - loss: 0.0013 - acc: 0.9996
Epoch 83/100
159491/159491 [==============================] - 31s 197us/sample - loss: 0.0013 - acc: 0.9996
Epoch 84/100
159491/159491 [==============================] - 31s 196us/sample - loss: 0.0016 - acc: 0.9996
Epoch 85/100
159491/159491 [==============================] - 31s 196us/sample - loss: 0.0012 - acc: 0.9996
Epoch 86/100
159491/159491 [==============================] - 31s 194us/sample - loss: 0.0013 - acc: 0.9997
Epoch 87/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0014 - acc: 0.9997
Epoch 88/100
159491/159491 [==============================] - 31s 195us/sample - loss: 0.0015 - acc: 0.9996
Epoch 89/100
159491/159491 [==============================] - 31s 194us/sample - loss: 0.0013 - acc: 0.9996
Epoch 90/100
159491/159491 [=============

In [33]:
feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

from tensorflow.keras.models import load_model

predict_list = np.zeros(39873);

n_members = 10
for i in range(n_members):
    model = load_model('model/model14x28x28x28x1_'+ str(i + 1)+ '.h5')
    
    #검증 값 저장
    predict = model.predict(test_X)
    predict_y = predict.reshape(-1)
    
    predict_list = predict_list + predict_y
    
predict_list = predict_list/10

predict = (predict_list > 0.5)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)




0.8652482269503547
0.8400423728813559


In [8]:
feature_total = feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

hidden_unit = 28
feature_count = 28

n_members = 10
for i in range(n_members):
    model = tf.keras.Sequential()
    model.add(layers.Dense(units=hidden_unit, 
                           kernel_initializer= keras.initializers.glorot_uniform(seed=None),
                           activation='relu',
                           input_dim = feature_count))

    model.add(layers.Dense(units=hidden_unit,
                           kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                           activation='relu'))

    model.add(layers.Dense(units=hidden_unit,
                           kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                           activation='relu'))

    model.add(layers.Dense(units=1,
                           kernel_initializer=keras.initializers.glorot_uniform(seed=None),
                           activation='sigmoid'))


    model.compile(optimizer='adam',
                  loss='binary_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_X, train_y,epochs=100,batch_size=32)
    
    filename = 'model/model28x28x28x28x1_'+ str(i + 1)+ '.h5'
    
    model.save(filename)

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Use tf.cast instead.
Epoch 1/100
159491/159491 [==============================] - 23s 146us/sample - loss: 0.0089 - acc: 0.9986
Epoch 2/100
159491/159491 [==============================] - 20s 125us/sample - loss: 0.0032 - acc: 0.9994
Epoch 3/100
159491/159491 [==============================] - 20s 125us/sample - loss: 0.0030 - acc: 0.9994
Epoch 4/100
159491/159491 [==============================] - 21s 131us/sample - loss: 0.0026 - acc: 0.9994
Epoch 5/100
159491/159491 [==============================] - 21s 129us/sample - loss: 0.0025 - acc: 0.9994
Epoch 6/100
159491/159491 [==============================] - 21s 130us/sample - loss: 0.0023 - acc: 0.9994
Epoch 7/100
159491/159491 [==============================] - 21s 132us/sample - loss: 0.0021 - acc: 0.9995
Epoch 8/100
159491/159491 [==============================] - 21s 131us/sample - loss: 0.0020 - acc: 0.9995
Epoch 9/100
159491/15949

159491/159491 [==============================] - 21s 132us/sample - loss: 3.8945e-04 - acc: 0.9999
Epoch 71/100
159491/159491 [==============================] - 21s 134us/sample - loss: 5.4502e-04 - acc: 0.9998
Epoch 72/100
159491/159491 [==============================] - 21s 132us/sample - loss: 2.9648e-04 - acc: 0.9999
Epoch 73/100
159491/159491 [==============================] - 21s 132us/sample - loss: 7.6167e-04 - acc: 0.9998
Epoch 74/100
159491/159491 [==============================] - 21s 133us/sample - loss: 6.6459e-04 - acc: 0.9998
Epoch 75/100
159491/159491 [==============================] - 21s 134us/sample - loss: 4.4808e-04 - acc: 0.9999
Epoch 76/100
159491/159491 [==============================] - 21s 132us/sample - loss: 7.2124e-04 - acc: 0.9998
Epoch 77/100
159491/159491 [==============================] - 21s 133us/sample - loss: 4.8111e-04 - acc: 0.9998
Epoch 78/100
159491/159491 [==============================] - 21s 133us/sample - loss: 5.5591e-04 - acc: 0.9999
Epoch

159491/159491 [==============================] - 21s 133us/sample - loss: 0.0011 - acc: 0.9997
Epoch 45/100
159491/159491 [==============================] - 22s 136us/sample - loss: 8.8652e-04 - acc: 0.9998
Epoch 46/100
159491/159491 [==============================] - 21s 131us/sample - loss: 7.8647e-04 - acc: 0.9997
Epoch 47/100
159491/159491 [==============================] - 21s 134us/sample - loss: 7.5477e-04 - acc: 0.9997
Epoch 48/100
159491/159491 [==============================] - 21s 134us/sample - loss: 6.7682e-04 - acc: 0.9998
Epoch 49/100
159491/159491 [==============================] - 21s 134us/sample - loss: 8.0556e-04 - acc: 0.9998
Epoch 50/100
159491/159491 [==============================] - 21s 133us/sample - loss: 0.0010 - acc: 0.9997
Epoch 51/100
159491/159491 [==============================] - 21s 134us/sample - loss: 7.6730e-04 - acc: 0.9997
Epoch 52/100
159491/159491 [==============================] - 21s 132us/sample - loss: 7.3956e-04 - acc: 0.9998
Epoch 53/100


159491/159491 [==============================] - 22s 135us/sample - loss: 0.0013 - acc: 0.9996
Epoch 19/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0013 - acc: 0.9996
Epoch 20/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0012 - acc: 0.9996
Epoch 21/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0012 - acc: 0.9996
Epoch 22/100
159491/159491 [==============================] - 21s 135us/sample - loss: 0.0011 - acc: 0.9997
Epoch 23/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0011 - acc: 0.9996
Epoch 24/100
159491/159491 [==============================] - 21s 134us/sample - loss: 0.0012 - acc: 0.9996
Epoch 25/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0011 - acc: 0.9996
Epoch 26/100
159491/159491 [==============================] - 21s 134us/sample - loss: 9.5115e-04 - acc: 0.9997
Epoch 27/100
159491/159491 [=========

159491/159491 [==============================] - 21s 133us/sample - loss: 3.5754e-04 - acc: 0.9999
Epoch 92/100
159491/159491 [==============================] - 22s 136us/sample - loss: 6.7346e-04 - acc: 0.9998
Epoch 93/100
159491/159491 [==============================] - 21s 133us/sample - loss: 4.4564e-04 - acc: 0.9998
Epoch 94/100
159491/159491 [==============================] - 21s 134us/sample - loss: 3.8271e-04 - acc: 0.9999
Epoch 95/100
159491/159491 [==============================] - 21s 134us/sample - loss: 5.5992e-04 - acc: 0.9998
Epoch 96/100
159491/159491 [==============================] - 21s 133us/sample - loss: 5.9350e-04 - acc: 0.9999
Epoch 97/100
159491/159491 [==============================] - 22s 136us/sample - loss: 5.6452e-04 - acc: 0.9998
Epoch 98/100
159491/159491 [==============================] - 21s 135us/sample - loss: 5.0727e-04 - acc: 0.9999
Epoch 99/100
159491/159491 [==============================] - 22s 137us/sample - loss: 3.0420e-04 - acc: 0.9999
Epoch

159491/159491 [==============================] - 22s 135us/sample - loss: 6.6122e-04 - acc: 0.9998
Epoch 66/100
159491/159491 [==============================] - 21s 135us/sample - loss: 6.1403e-04 - acc: 0.9998
Epoch 67/100
159491/159491 [==============================] - 21s 133us/sample - loss: 5.5641e-04 - acc: 0.9997
Epoch 68/100
159491/159491 [==============================] - 22s 136us/sample - loss: 6.4633e-04 - acc: 0.9998
Epoch 69/100
159491/159491 [==============================] - 21s 135us/sample - loss: 7.0225e-04 - acc: 0.9998
Epoch 70/100
159491/159491 [==============================] - 22s 135us/sample - loss: 7.6833e-04 - acc: 0.9998
Epoch 71/100
159491/159491 [==============================] - 22s 138us/sample - loss: 5.9055e-04 - acc: 0.9998
Epoch 72/100
159491/159491 [==============================] - 22s 135us/sample - loss: 7.3831e-04 - acc: 0.9997
Epoch 73/100
159491/159491 [==============================] - 21s 135us/sample - loss: 5.9623e-04 - acc: 0.9998
Epoch

159491/159491 [==============================] - 22s 135us/sample - loss: 8.5501e-04 - acc: 0.9998
Epoch 40/100
159491/159491 [==============================] - 22s 135us/sample - loss: 9.4958e-04 - acc: 0.9997
Epoch 41/100
159491/159491 [==============================] - 22s 136us/sample - loss: 7.7714e-04 - acc: 0.9997
Epoch 42/100
159491/159491 [==============================] - 22s 138us/sample - loss: 7.7701e-04 - acc: 0.9997
Epoch 43/100
159491/159491 [==============================] - 22s 136us/sample - loss: 7.6303e-04 - acc: 0.9997
Epoch 44/100
159491/159491 [==============================] - 22s 136us/sample - loss: 7.6549e-04 - acc: 0.9997
Epoch 45/100
159491/159491 [==============================] - 22s 137us/sample - loss: 7.6910e-04 - acc: 0.9997
Epoch 46/100
159491/159491 [==============================] - 22s 138us/sample - loss: 7.5248e-04 - acc: 0.9997
Epoch 47/100
159491/159491 [==============================] - 21s 134us/sample - loss: 7.1627e-04 - acc: 0.9998
Epoch

159491/159491 [==============================] - 21s 134us/sample - loss: 0.0017 - acc: 0.9996
Epoch 13/100
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0017 - acc: 0.9995
Epoch 14/100
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0015 - acc: 0.9996
Epoch 15/100
159491/159491 [==============================] - 22s 135us/sample - loss: 0.0014 - acc: 0.9996
Epoch 16/100
159491/159491 [==============================] - 22s 138us/sample - loss: 0.0016 - acc: 0.9996
Epoch 17/100
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0014 - acc: 0.9996
Epoch 18/100
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0014 - acc: 0.9996
Epoch 19/100
159491/159491 [==============================] - 22s 139us/sample - loss: 0.0013 - acc: 0.9996
Epoch 20/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0012 - acc: 0.9997
Epoch 21/100
159491/159491 [=============

Epoch 86/100
159491/159491 [==============================] - 21s 134us/sample - loss: 3.2925e-04 - acc: 0.9999
Epoch 87/100
159491/159491 [==============================] - 21s 132us/sample - loss: 5.3314e-04 - acc: 0.9998
Epoch 88/100
159491/159491 [==============================] - 21s 131us/sample - loss: 3.8952e-04 - acc: 0.9999
Epoch 89/100
159491/159491 [==============================] - 21s 134us/sample - loss: 5.3149e-04 - acc: 0.9998
Epoch 90/100
159491/159491 [==============================] - 22s 137us/sample - loss: 5.8963e-04 - acc: 0.9999
Epoch 91/100
159491/159491 [==============================] - 22s 136us/sample - loss: 2.9110e-04 - acc: 0.9999
Epoch 92/100
159491/159491 [==============================] - 22s 135us/sample - loss: 5.8434e-04 - acc: 0.9999
Epoch 93/100
159491/159491 [==============================] - 22s 137us/sample - loss: 4.8366e-04 - acc: 0.9999
Epoch 94/100
159491/159491 [==============================] - 22s 136us/sample - loss: 4.0305e-04 - acc:

159491/159491 [==============================] - 21s 133us/sample - loss: 9.4032e-04 - acc: 0.9998
Epoch 61/100
159491/159491 [==============================] - 22s 136us/sample - loss: 8.9920e-04 - acc: 0.9998
Epoch 62/100
159491/159491 [==============================] - 21s 134us/sample - loss: 7.2725e-04 - acc: 0.9998
Epoch 63/100
159491/159491 [==============================] - 22s 137us/sample - loss: 5.3680e-04 - acc: 0.9998
Epoch 64/100
159491/159491 [==============================] - 22s 136us/sample - loss: 7.3809e-04 - acc: 0.9998
Epoch 65/100
159491/159491 [==============================] - 22s 138us/sample - loss: 4.9578e-04 - acc: 0.9998
Epoch 66/100
159491/159491 [==============================] - 22s 139us/sample - loss: 5.9748e-04 - acc: 0.9998
Epoch 67/100
159491/159491 [==============================] - 22s 135us/sample - loss: 4.3140e-04 - acc: 0.9999
Epoch 68/100
159491/159491 [==============================] - 21s 133us/sample - loss: 6.5533e-04 - acc: 0.9998
Epoch

159491/159491 [==============================] - 22s 138us/sample - loss: 0.0010 - acc: 0.9998
Epoch 35/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0011 - acc: 0.9997
Epoch 36/100
159491/159491 [==============================] - 22s 138us/sample - loss: 7.1805e-04 - acc: 0.9998
Epoch 37/100
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0013 - acc: 0.9997
Epoch 38/100
159491/159491 [==============================] - 22s 137us/sample - loss: 8.4381e-04 - acc: 0.9997
Epoch 39/100
159491/159491 [==============================] - 22s 137us/sample - loss: 9.5777e-04 - acc: 0.9997
Epoch 40/100
159491/159491 [==============================] - 22s 141us/sample - loss: 7.3733e-04 - acc: 0.9998
Epoch 41/100
159491/159491 [==============================] - 22s 141us/sample - loss: 8.4911e-04 - acc: 0.9997
Epoch 42/100
159491/159491 [==============================] - 22s 137us/sample - loss: 8.8650e-04 - acc: 0.9998
Epoch 43/100
1594

159491/159491 [==============================] - 22s 138us/sample - loss: 0.0023 - acc: 0.9995
Epoch 8/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0022 - acc: 0.9995
Epoch 9/100
159491/159491 [==============================] - 22s 138us/sample - loss: 0.0021 - acc: 0.9995
Epoch 10/100
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0019 - acc: 0.9995
Epoch 11/100
159491/159491 [==============================] - 22s 140us/sample - loss: 0.0019 - acc: 0.9995
Epoch 12/100
159491/159491 [==============================] - 22s 138us/sample - loss: 0.0017 - acc: 0.9995
Epoch 13/100
159491/159491 [==============================] - 22s 136us/sample - loss: 0.0018 - acc: 0.9995
Epoch 14/100
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0017 - acc: 0.9996
Epoch 15/100
159491/159491 [==============================] - 22s 137us/sample - loss: 0.0017 - acc: 0.9995
Epoch 16/100
159491/159491 [===============

159491/159491 [==============================] - 22s 141us/sample - loss: 8.5042e-04 - acc: 0.9998
Epoch 82/100
159491/159491 [==============================] - 22s 139us/sample - loss: 8.3742e-04 - acc: 0.9998
Epoch 83/100
159491/159491 [==============================] - 22s 136us/sample - loss: 4.5717e-04 - acc: 0.9999
Epoch 84/100
159491/159491 [==============================] - 22s 137us/sample - loss: 5.6979e-04 - acc: 0.9998
Epoch 85/100
159491/159491 [==============================] - 22s 138us/sample - loss: 6.9586e-04 - acc: 0.9998
Epoch 86/100
159491/159491 [==============================] - 22s 139us/sample - loss: 4.7395e-04 - acc: 0.9999
Epoch 87/100
159491/159491 [==============================] - 22s 138us/sample - loss: 7.4346e-04 - acc: 0.9999
Epoch 88/100
159491/159491 [==============================] - 22s 139us/sample - loss: 5.3311e-04 - acc: 0.9999
Epoch 89/100
159491/159491 [==============================] - 22s 139us/sample - loss: 4.5181e-04 - acc: 0.9999
Epoch

159491/159491 [==============================] - 22s 138us/sample - loss: 4.9650e-04 - acc: 0.9998
Epoch 56/100
159491/159491 [==============================] - 22s 139us/sample - loss: 6.0222e-04 - acc: 0.9998
Epoch 57/100
159491/159491 [==============================] - 22s 136us/sample - loss: 6.9876e-04 - acc: 0.9998
Epoch 58/100
159491/159491 [==============================] - 22s 137us/sample - loss: 5.4183e-04 - acc: 0.9998
Epoch 59/100
159491/159491 [==============================] - 22s 139us/sample - loss: 6.8864e-04 - acc: 0.9998
Epoch 60/100
159491/159491 [==============================] - 22s 136us/sample - loss: 6.1731e-04 - acc: 0.9998
Epoch 61/100
159491/159491 [==============================] - 22s 140us/sample - loss: 4.2801e-04 - acc: 0.9999
Epoch 62/100
159491/159491 [==============================] - 22s 137us/sample - loss: 5.4749e-04 - acc: 0.9998
Epoch 63/100
159491/159491 [==============================] - 22s 141us/sample - loss: 5.2171e-04 - acc: 0.9999
Epoch

In [32]:
feature_total = feature_total = ['V1','V2','V3','V4','V5','V6','V7','V8','V9','V10','V11','V12','V13','V14','V15',
              'V16','V17','V18','V19','V20','V21','V22','V23','V24','V25','V26','V27','V28'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

from tensorflow.keras.models import load_model

predict_list = np.zeros(39873);

n_members = 10
for i in range(n_members):
    model = load_model('model/model28x28x28x28x1_'+ str(i + 1)+ '.h5')
    
    #검증 값 저장
    predict = model.predict(test_X)
    predict_y = predict.reshape(-1)
    
    predict_list = predict_list + predict_y
    
predict_list = predict_list/10

predict = (predict_list > 0.5)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)


0.838235294117647
0.801948051948052


# 결론

가장 좋은 학습 값은 Neural Network 3 Hidden Layer 10 time Ensemble (with feature 14)로 나왔다

앙상블은 feature14에 대해 100epoch을 트레이닝한 model 10개의 결과를 평균내어 사용하였다

In [78]:
feature_total = ['V1','V2','V3','V4','V5','V7','V9','V10','V11','V12','V14','V16','V17','V18'];

train_X,train_y = SelectFeature(train_set,feature_total);
test_X , test_y = SelectFeature(test_set,feature_total);

from tensorflow.keras.models import load_model

predict_list = np.zeros(39873);

n_members = 10
for i in range(n_members):
    model = load_model('model/model14x28x28x28x1_'+ str(i + 1)+ '.h5')
    
    #검증 값 저장
    predict = model.predict(test_X)
    predict_y = predict.reshape(-1)
    
    predict_list = predict_list + predict_y
    
predict_list = predict_list/10

predict = (predict_list > 0.5)
CalScore(test_y,predict)
CalBetaScore(test_y,predict)



0.8652482269503547
0.8400423728813559


모델의 예측 확률 값과 예측 값

In [79]:
#확률값
print(predict_list)
#예측 Class
print(predict)

[0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 7.23809004e-05
 0.00000000e+00 8.97165058e-20]
[False False False ... False False False]


In [80]:
series = pd.Series(predict_list)
print(series)

0        0.000000e+00
1        0.000000e+00
2        0.000000e+00
3        0.000000e+00
4        0.000000e+00
5        0.000000e+00
6        0.000000e+00
7        0.000000e+00
8        1.490116e-08
9        0.000000e+00
10       0.000000e+00
11       0.000000e+00
12       0.000000e+00
13       2.950430e-07
14       1.192093e-08
15       3.676713e-05
16       0.000000e+00
17       4.142523e-07
18       0.000000e+00
19       0.000000e+00
20       0.000000e+00
21       5.960464e-09
22       0.000000e+00
23       0.000000e+00
24       0.000000e+00
25       0.000000e+00
26       0.000000e+00
27       0.000000e+00
28       0.000000e+00
29       0.000000e+00
             ...     
39843    0.000000e+00
39844    0.000000e+00
39845    0.000000e+00
39846    6.854534e-08
39847    0.000000e+00
39848    0.000000e+00
39849    0.000000e+00
39850    0.000000e+00
39851    0.000000e+00
39852    1.519918e-07
39853    0.000000e+00
39854    0.000000e+00
39855    0.000000e+00
39856    0.000000e+00
39857    0

In [81]:
#랭킹
rankSeries = series.rank(method='first',ascending=False)
rank = rankSeries.astype('int32').values
print(rank)

[ 7812  7813  7814 ...  1337 39873  7811]


Neural Network를 사용하였을 때 가장 높은 값이 나왔지만, Random Forest같은 경우도 f1:0.842857143, fbeta:0.81595744680851라는 상당한 높은 수치가 나왔으므로 하이퍼 파라미터를 적절히 튜닝한다면 NN급의 성능을 보일 수 있을 것 같다